New packages: Hunspell dictionaries for English, Spanish, Italian and French.

k0001 opened the pull request:
These aren't readily picked up by Hunspell-compatible software such as Firefox or LibreOffice just yet, as those programs need to be setup to look for the dictionaries at the proper paths, which involves changing the derivations for those programs. Nevertheless, one can point Hunspell-compatible tools to, say, `$HOME/.nix-profile/share/hunspell` and they will find the dictionaries:

```
% echo "bonjor" | hunspell -d fr-classique
Can't open affix or dictionary files for dictionary named "fr-classique".

% export DICPATH=$HOME/.nix-profile/share/hunspell
% echo "bonjor" | hunspell -d fr-classique
Hunspell 1.3.3
& bonjor 1 0: bonjour
```

The full list of packages added:

```
hunspellDicts.es-any         hunspell-dict-es-any-rediris-0.7
hunspellDicts.es-ar          hunspell-dict-es-ar-rediris-0.7
hunspellDicts.es-bo          hunspell-dict-es-bo-rediris-0.7
hunspellDicts.es-cl          hunspell-dict-es-cl-rediris-0.7
hunspellDicts.es-co          hunspell-dict-es-co-rediris-0.7
hunspellDicts.es-cr          hunspell-dict-es-cr-rediris-0.7
hunspellDicts.es-cu          hunspell-dict-es-cu-rediris-0.7
hunspellDicts.es-do          hunspell-dict-es-do-rediris-0.7
hunspellDicts.es-ec          hunspell-dict-es-ec-rediris-0.7
hunspellDicts.es-es          hunspell-dict-es-es-rediris-0.7
hunspellDicts.es-gt          hunspell-dict-es-gt-rediris-0.7
hunspellDicts.es-hn          hunspell-dict-es-hn-rediris-0.7
hunspellDicts.es-mx          hunspell-dict-es-mx-rediris-0.7
hunspellDicts.es-ni          hunspell-dict-es-ni-rediris-0.7
hunspellDicts.es-pa          hunspell-dict-es-pa-rediris-0.7
hunspellDicts.es-pe          hunspell-dict-es-pe-rediris-0.7
hunspellDicts.es-pr          hunspell-dict-es-pr-rediris-0.7
hunspellDicts.es-py          hunspell-dict-es-py-rediris-0.7
hunspellDicts.es-sv          hunspell-dict-es-sv-rediris-0.7
hunspellDicts.es-uy          hunspell-dict-es-uy-rediris-0.7
hunspellDicts.es-ve          hunspell-dict-es-ve-rediris-0.7
hunspellDicts.fr-any         hunspell-dict-fr-any-dicollecte-5.2
hunspellDicts.fr-classique   hunspell-dict-fr-classique-dicollecte-5.2
hunspellDicts.fr-moderne     hunspell-dict-fr-moderne-dicollecte-5.2
hunspellDicts.fr-reforme1990 hunspell-dict-fr-reforme1990-dicollecte-5.2
hunspellDicts.it-it          hunspell-dict-it-it-linguistico-2.4
```

Hint: in Firefox you can go to `about:config` and create a new string value named `spellchecker.dictionary_path` set to `PATH-TO-YOUR-HOME-FOLDER/.nix-profile/share/hunspell` and the dictionaries will be available after restarting Firefox.

k0001 committed:
New packages: Hunspell dictionaries for Spanish, Italian and French.

These aren't readily picked up by Hunspell-compatible software such as
Firefox or LibreOffice just yet, as they need to be setup to look for
the dictionaries at the proper paths. Nevertheless, one can point
Hunspell-compatible tools to, say, $HOME/.nix-profile/share/hunspell and
they will find the dictionaries:

  % echo "bonjor" | hunspell -d fr-classique
  Can't open affix or dictionary files for dictionary named "fr-classique".

  % export DICPATH=$HOME/.nix-profile/share/hunspell
  % echo "bonjor" | hunspell -d fr-classique
  Hunspell 1.3.3
  & bonjor 1 0: bonjour

The full list of packages added:

hunspellDicts.es-any         hunspell-dict-es-any-rediris-0.7
hunspellDicts.es-ar          hunspell-dict-es-ar-rediris-0.7
hunspellDicts.es-bo          hunspell-dict-es-bo-rediris-0.7
hunspellDicts.es-cl          hunspell-dict-es-cl-rediris-0.7
hunspellDicts.es-co          hunspell-dict-es-co-rediris-0.7
hunspellDicts.es-cr          hunspell-dict-es-cr-rediris-0.7
hunspellDicts.es-cu          hunspell-dict-es-cu-rediris-0.7
hunspellDicts.es-do          hunspell-dict-es-do-rediris-0.7
hunspellDicts.es-ec          hunspell-dict-es-ec-rediris-0.7
hunspellDicts.es-es          hunspell-dict-es-es-rediris-0.7
hunspellDicts.es-gt          hunspell-dict-es-gt-rediris-0.7
hunspellDicts.es-hn          hunspell-dict-es-hn-rediris-0.7
hunspellDicts.es-mx          hunspell-dict-es-mx-rediris-0.7
hunspellDicts.es-ni          hunspell-dict-es-ni-rediris-0.7
hunspellDicts.es-pa          hunspell-dict-es-pa-rediris-0.7
hunspellDicts.es-pe          hunspell-dict-es-pe-rediris-0.7
hunspellDicts.es-pr          hunspell-dict-es-pr-rediris-0.7
hunspellDicts.es-py          hunspell-dict-es-py-rediris-0.7
hunspellDicts.es-sv          hunspell-dict-es-sv-rediris-0.7
hunspellDicts.es-uy          hunspell-dict-es-uy-rediris-0.7
hunspellDicts.es-ve          hunspell-dict-es-ve-rediris-0.7
hunspellDicts.fr-any         hunspell-dict-fr-any-dicollecte-5.2
hunspellDicts.fr-classique   hunspell-dict-fr-classique-dicollecte-5.2
hunspellDicts.fr-moderne     hunspell-dict-fr-moderne-dicollecte-5.2
hunspellDicts.fr-reforme1990 hunspell-dict-fr-reforme1990-dicollecte-5.2
hunspellDicts.it-it          hunspell-dict-it-it-linguistico-2.4

edwtjo commented on pkgs/development/libraries/hunspell/dictionaries.nix:
```diff
@@ -0,0 +1,368 @@
+      inherit src;
+      version = "0.7";
+      name = "hunspell-dict-${shortName}-rediris-${version}";
+      meta = {
```
+ You can avoid some typing using, `meta = with stdenv.lib; {`
+ No maintainer field

edwtjo commented:
Looks good, works as advertised.

k0001 committed:
Add myself to maintainers.nix

k0001 committed:
Clean up nixpkgs/development/libraries/hunspell/dictionaries.nix

k0001 committed:
New packages: English Hunspell dictionaries

hunspellDicts.en-us      hunspell-dict-en-us-wordlist-2014.11.17
hunspellDicts.en-ca      hunspell-dict-en-ca-wordlist-2014.11.17
hunspellDicts.en-gb-ise  hunspell-dict-en-gb-ise-wordlist-2014.11.17
hunspellDicts.en-gb-ize  hunspell-dict-en-gb-ize-wordlist-2014.11.17

k0001 commented:
I went ahead and added English dictionaries too. There's a bit of repetition in the `installPhase` of some derivations, but haven't figured out a proper way to abstract that away. I'm still not that confortable with writing package derivations.

edwtjo commented:
Ok, so I took the liberty of committing a refactored version. Also, everybody does things a bit differently. I prefer using local specialised attributes and functions instead of using override. Note though that the ... can be translated into: "there might be other stuff in the set but I don't care". you could add a @anyName or anyName@ in order to name the set. This can for instance be used in conjunction with ...  and override() to manipulate its arguments.

Closed with 886575ec1a51f0756c7929f24183fcef594491f5

k0001 commented:
Thanks @edwtjo, I have a better understanding now of how these things are supposed to be packaged. 

I noticed, though, that some details about the Italian dictionary leaked into the general `mkDict`. I guess it is not that bad, as these derivations are quite self contained within this single file and no one will care. But perhaps next time me or somebody else adds a dictionary, that could be cleaned up too. 

Thanks for your hints, and for taking the time to improve and merge this :)

alexeymuranov commented:
I installed a few `hunspell` dictionaries and observed the following phenomenon: one of the dictionaries, `hunspell-dict-en-us-wordlist-2014.11.17`, was immediately picked up by *Evolution* e-mail client and by *gedit* text editor (and probably by everything else), but none of the other dictionaries that i installed was recognised.

What is going on?  Can other dictionaries be fixed to get recognised by the applications?  (If so, I might be ready to look into it.)

I also noticed that i have Hebrew spelling available even though there is no dictionary explicitly installed for it.  Strange.
